dotnet/src/Experimental/Process.LocalRuntime/LocalStep.cs

```diff
@@ -0,0 +1,274 @@
+
+        // Instantiate an instance of the inner step object
+        KernelProcessStep stepInstance = (KernelProcessStep)ActivatorUtilities.CreateInstance(this._kernel.Services, stepInfo.InnerStepType);
+        var kernelPlugin = KernelPluginFactory.CreateFromObject(stepInstance);
+
+        // Load the kernel functions
+        foreach (KernelFunction f in kernelPlugin)
+        {
+            Verify.NotNull(f);
+            var cloned = f.Clone(stepInfo.State.Name!);
+            this._functions.Add(cloned.Name, cloned);
+        }
+
+        // Initialize the input channels
+        this._initialInputs = this.FindInputChannels();
+        this._inputs = new(this._initialInputs);
+
+        // Activate the step with user-defined state if needed
+        KernelProcessStepState? stateObject = null;
+        Type? stateType = null;
+
+        if (TryGetSubtypeOfStatefulStep(stepInfo.InnerStepType, out Type? genericStepType) && genericStepType is not null)
+        {
+            // The step is a subclass of KernelProcessStep<>, so we need to extract the generic type argument
+            // and create an instance of the corresponding KernelProcessStepState<>.
+            var userStateType = genericStepType.GetGenericArguments()[0];
+            Verify.NotNull(userStateType);
+
+            stateType = typeof(KernelProcessStepState<>).MakeGenericType(userStateType);
+            Verify.NotNull(stateType);
+
+            stateObject = (KernelProcessStepState?)Activator.CreateInstance(stateType, this.Id, this.Name);
+        }
+        else
+        {
+            // The step is a KernelProcessStep with no user-defined state, so we can use the base KernelProcessStepState.
+            stateType = typeof(KernelProcessStepState);
+            stateObject = new KernelProcessStepState(this.Id, this.Name);
+        }
+
+        Verify.NotNull(stateObject);
+        MethodInfo? methodInfo = stepInfo.InnerStepType.GetMethod(nameof(KernelProcessStep.ActivateAsync), [stateType]);
+        Verify.NotNull(methodInfo);
+
+        methodInfo.Invoke(stepInstance, [stateObject]);
+        await stepInstance.ActivateAsync(stateObject).ConfigureAwait(false);
+    }
+
+    public IEnumerable<KernelProcessEvent> GetAllEvents()
+    {
+        var allEvents = this._eventQueue.ToList();
+        this._eventQueue.Clear();
+        return allEvents;
+    }
+
+    public List<KernelProcessEdge> GetEdgeForEvent(string eventId)
+    {
+        if (this._outputEdges is null)
+        {
+            return [];
+        }
+
+        if (this._outputEdges.TryGetValue(eventId, out List<KernelProcessEdge>? edges) && edges is not null)
+        {
+            return edges;
+        }
+
+        return [];
+    }
+
+    public async Task HandleMessageAsync(LocalMessage message)
+    {
+        if (this._functions is null || this._inputs is null || this._initialInputs is null)
+        {
+            throw new InvalidOperationException("The step has not been initialized.");
```

Comment:
I might strongly consider introducing a `ProcessException` to that exceptions specific to this framework might be clearly distinguishable from other system classes.

Reply:
I'll use KernelException for now and create a work item to determine if a process specific exception is useful.